controllers/openstacknet_controller.go

```diff
@@ -93,6 +95,48 @@ func (r *OpenStackNetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, err
 	}
 
+	// examine DeletionTimestamp to determine if object is under deletion
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(instance, openstacknet.FinalizerName) {
+			controllerutil.AddFinalizer(instance, openstacknet.FinalizerName)
+			if err := r.Update(context.Background(), instance); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Finalizer %s added to CR %s", openstacknet.FinalizerName, instance.Name))
+		}
+	} else {
+		// 1. check if finalizer is there
+		// Reconcile if finalizer got already removed
+		if !controllerutil.ContainsFinalizer(instance, openstacknet.FinalizerName) {
+			return ctrl.Result{}, nil
+		}
+
+		// 2. Clean up resources used by the operator
+		// SRIOV resources
+		err = r.sriovResourceCleanup(instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// 3. as last step remove the finalizer on the operator CR to finish delete
+		controllerutil.RemoveFinalizer(instance, openstacknet.FinalizerName)
+		err = r.Client.Update(context.TODO(), instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		r.Log.Info(fmt.Sprintf("CR %s deleted", instance.Name))
+		return ctrl.Result{}, nil
+	}
+
+	if instance.Spec.AttachConfiguration.NodeSriovConfigurationPolicy.DesiredState.Port != "" {
+		if err := r.ensureSriov(instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
 	// generate NodeNetworkConfigurationPolicy
```

Comment:
we probably need a check here to skip non sriov network configuration if there is no instance.Spec.AttachConfiguration.NodeNetworkConfigurationPolicy ?

Reply:
Yep, forgot the `else` :)